Review of the Getting Started content (#404)
neo4j/docs-getting-started: modules/ROOT/pages/tutorials/relational-to-graph-modeling.adoc

[[relational-to-graph-modeling]]
= Modeling: relational to graph
:tags: graph-modeling, data-model, schema, relational-to-graph, translating-model
:description: For those with a background in relational data modeling, this guide helps transfer your existing knowledge of the processes and components used for relational data modeling into graph data modeling.

> [abstract]

Reviewer: Do we use abstracts still?

{description}

> It will help to compare and contrast the steps of each process and help you identify where the data modeling is similar or different for each type of database.

Reviewer: ```suggestion
It is helpful to compare and contrast the steps of each process to help you identify where data modeling is similar or different for each type of database.
```

[#rdbms-to-graph-model]
== Introduction

If you are familiar with the relational data model that has tables, columns, relationship cardinalities, and other components, graph data modeling will not seem entirely foreign.
The design of the data model still needs to be based upon requirements for access, queries, performance expectation, and business logic.
However, the structure of a graph data model is laid out slightly differently.

> You may have an entirely new project that you want to create a graph data model, but are only familiar with how to create the relational model.

Reviewer: ```suggestion
You may have an entirely new project in which you want to create a graph data model but are only familiar with how to create the relational model.
```

Or you could already have an existing project with a relational model that you want to convert to graph.
Either way, this guide will take existing knowledge of the relational data model and show you how to use that to create a graph model.

[#rdbms-graph-architecture]
== Relational and graph architecture

As a quick overview, remember that relational databases rely upon index lookups and table joins to connect different entities.
This quickly becomes a problem for performance, especially when there are several tables joined, millions of rows on tables, or complex queries that traverse various levels through subqueries.

> In our example from the concepts page, to find which departments Alice works for, you would need to query the `Person` table to find the row representing Alice, which is tied to a unique ID as the primary key.

Reviewer: ```suggestion
In the example from the concepts page, to find which departments Alice works for, you would need to query the `Person` table to find the row representing Alice, which is tied to a unique ID as the primary key.
```
Make sure that this example is actually used.

Then, your query would go to the associative entity table (`Person_Dept`) to find where her ID is tied to one or more department IDs.
Finally, the query would check the `Department` table to find the actual values for those department IDs you found in the associative entity table.

> The image below reviews this example we just described.

Reviewer: ```suggestion
The image below reviews this example.
```

.Relational - Person and Department tables
image::relational_model.svg[role="popup-link",width=600]

> In a graph, you do not need to worry about table joins and index lookups because graph data is structured by each, individual entity and its relationships with other individual entities.

Reviewer: ```suggestion
In a graph, you do not need to worry about table joins and index lookups because graph data is structured by each, individual entity and its relationships with other individual entities.
```
```suggestion
In a graph, you do not need to consider table joins and index lookups because graph data is structured by each individual entity and its relationships with other individual entities.
```

> Ok, so how do we go from creating relational data models to a graph data model?

Reviewer: ```suggestion
How do you go from creating relational data models to a graph data model?
```

[#model-transformation]
== Data model transformation tips

> Let us look at some of the key components in a relational data model and translate those into components of a graph data model.

Reviewer: ```suggestion
Have a look at some of the key components in a relational data model and translate those into components of a graph data model.
```

- *_Table to Node Label_* - each entity table in the relational model becomes a label on nodes in the graph model.
- *_Row to Node_* - each row in a relational entity table becomes a node in the graph.
- *_Column to Node Property_* - columns (fields) on the relational tables become node properties in the graph.
- *_Business primary keys only_* - remove technical primary keys, keep business primary keys.
- *_Add Constraints/Indexes_* - add unique constraints for business primary keys, add indexes for frequent lookup attributes.
- *_Foreign keys to Relationships_* - replace foreign keys to the other table with relationships, remove them afterwards.
- *_No defaults_* - remove data with default values, no need to store those.
- *_Clean up data_* - duplicate data in denormalized tables might have to be pulled out into separate nodes to get a cleaner model.
- *_Index Columns to Array_* - indexed column names (like email1, email2, email3) might indicate an array property.
- *_Join tables to Relationships_* - join tables are transformed into relationships, columns on those tables become relationship properties

> If you apply the items in the list above to our example finding Alice's departments, we can come to a graph like the one shown below.

Reviewer: ```suggestion
If you apply the items in the list above to the example finding Alice's departments, you should end up with a graph like the following:
```

.Graph - Alice and three departments as nodes
image::relational_graph_model-arr.svg[role="popup-link",width=350]

Though the two models have similarities such as categorizing data by using either a table structure or a label, the graph model does not confine data to a pre-defined and strict table/column layout.

> We will look at another example in the next section.

Reviewer: ```suggestion

```

[#org-domain-model]
== Organizational domain data model

> To give us another chance to practice, we will use a standard organizational domain and show how it would be modeled in a relational database versus a graph database. To give yourself an extra challenge, try to create the graph data model on your own and then see how closely it lines up.

Reviewer: ```suggestion
To give you another chance to practice, you will see a standard organizational domain and see how it would be modeled in a relational database versus a graph database. 
To give yourself an extra challenge, try to create the graph data model on your own and then see how closely it lines up.
```

.Organizational domain - Relational model
image::relational_org_chart.svg[role="popup-link",width=600]

=== Conversion steps

> First, we can categorize our tables by main domain tables and associative entity tables by colors.
> Then, we can turn our table names into node labels.
> In this case, `Project`, `Person`, `Department`, and `Organization` become labels in our graph model.
> 
> The rows on our tables become their own nodes and the columns in those rows become the properties on those nodes.
> For example, your row on the `Person` table will become a node with your name and date of birth as the properties on your node.
> Any indexed columns that allow multiple similar values will become an array (such as skill1, skill2, skill3 columns translate to three values stored in an array property on a node).

Reviewer: ```suggestion
First, you can categorize your tables by main domain tables and associative entity tables by colors.
Then, you can turn your table names into node labels.
In this case, `Project`, `Person`, `Department`, and `Organization` become labels in your graph model.

The rows on your tables become their own nodes and the columns in those rows become the properties on those nodes.
For example, your row on the `Person` table will become a node with your name and date of birth as the properties on your node.
Any indexed columns that allow multiple similar values will become an array (such as skill1, skill2, skill3 columns translate to three values stored in an array property on a node).
```

You will also need to add unique constraints for the business primary keys in order to ensure the database will not allow duplicates.

Foreign keys that would aid in relational join lookups are transformed into relationships, as they show the links between the nodes.
Join tables (or associative entity tables) become relationships, as well, with any join table columns moved to relationship properties.

Since you only store the needed properties in Neo4j, you do not need to store nulls and empty values, so you can remove any default values that may have been created in a relational model.

Finally, any duplicate data created to normalize tables or de-normalize for simplicity's sake needs removed, as it is unneeded in a graph.

After this process, your graph data model should look something like the image below.

.Organizational domain - Graph model
image::graph_org_chart-arr.svg[width=500,role="popup-link"]

> It is important to have an basic understanding of the graph model before you start to import data, as it becomes easier to hydrate that model or adjust it later, as needs change.

Reviewer: hydrate?

> In an upcoming guide, how you model your graph data can impact queries, performance, and model changes.

Reviewer: ```suggestion
In the [insert name and link to this guide here] guide, you will learn how the graph date is modeled can impact queries, performance, and model changes.
```
